src/main/kotlin/org/rust/ide/annotator/RsFormatMacroAnnotator.kt

```diff
@@ -0,0 +1,127 @@
+
+import com.intellij.ide.annotator.AnnotatorBase
+import com.intellij.lang.annotation.AnnotationHolder
+import com.intellij.lang.annotation.HighlightSeverity
+import com.intellij.openapi.util.TextRange
+import com.intellij.openapiext.isUnitTestMode
+import com.intellij.psi.PsiElement
+import org.rust.ide.colors.RsColor
+import org.rust.lang.core.psi.RsLitExpr
+import org.rust.lang.core.psi.RsLiteralKind
+import org.rust.lang.core.psi.RsMacroCall
+import org.rust.lang.core.psi.ext.descendantOfTypeStrict
+import org.rust.lang.core.psi.ext.macroName
+import org.rust.lang.core.psi.ext.startOffset
+import org.rust.lang.core.psi.kind
+import org.rust.lang.utils.parseRustStringCharacters
+
+class RsFormatMacroAnnotator : AnnotatorBase() {
+
+    override fun annotateInternal(element: PsiElement, holder: AnnotationHolder) {
+        val formatMacro = element as? RsMacroCall ?: return
+
+        val formatStr = formatMacro
+            .formatMacroArgument
+            ?.formatMacroArgList
+            ?.getOrNull(macroToFormatPos(formatMacro.macroName))
+            ?.descendantOfTypeStrict<RsLitExpr>()
+            ?: return
+
+        val literalKind = (formatStr.kind as? RsLiteralKind.String) ?: return
+        val rawTextRange = literalKind.offsets.value ?: return
+        val (unescapedText, sourceMap, _) = parseRustStringCharacters(rawTextRange.substring(literalKind.node.text))
+
+        val arguments = formatParser.findAll(unescapedText.toString())
+        var currentOffset = 0
+
+        fun IntRange.toSourceRange(additionalOffset: Int = currentOffset): TextRange =
+            TextRange(sourceMap[first + additionalOffset], sourceMap[last + additionalOffset] + 1)
+                .shiftRight(formatStr.startOffset + rawTextRange.startOffset)
+
+        fun AnnotationHolder.highlightArgument(range: IntRange?, color: RsColor = RsColor.IDENTIFIER) {
+            if (range != null && !range.isEmpty())
+                createAnnotation(
+                    ARG_HIGHLIGHTING,
+                    range.toSourceRange(),
+                    null)
+                    .textAttributes = color.textAttributesKey
+        }
+
+        val key = RsColor.VALID_STRING_ESCAPE
+        val highlightSeverity = if (isUnitTestMode) key.testSeverity else HighlightSeverity.INFORMATION
+        for (arg in arguments) {
+            currentOffset = 0
+            val textRange = arg.range.toSourceRange()
+            holder.createAnnotation(highlightSeverity, textRange, null)
+                .textAttributes = key.textAttributesKey
+
+            if (arg.groups[3] != null) {
+                holder.createErrorAnnotation(
+                    arg.range.toSourceRange(),
+                    "Invalid format string: unmatched '}'")
+            }
+            if (arg.groups[1] != null) {
+
+                val inside = arg.groups[2]!!
+                currentOffset = inside.range.first
+                val matchResult = formatArgParser.find(inside.value)!!
```

Comment:
Hmm, let's add a comment about why this can't be null. Or even better pattern is `?: error("This can't be null because ...")`